src/librustc/sir.rs

```diff
@@ -1,174 +1,112 @@
 //! Serialised Intermediate Representation (SIR).
 //!
-//! SIR is built in-memory during LLVM code-generation, and finally placed into a dedicated ELF
-//! section at link time.
-
-use rustc_data_structures::fx::FxHashMap;
-use rustc_index::{newtype_index, vec::IndexVec};
+//! SIR is built in-memory during code-generation (in rustc_codegen_ssa), and finally placed
+//! into an ELF section at link time.
+
+use crate::ty::{Instance, TyCtxt};
+use rustc::mir;
+use rustc_hir::def_id::LOCAL_CRATE;
+use rustc_session::config::OutputType;
+use rustc_span::sym;
+use std::cell::RefCell;
+use std::convert::TryFrom;
 use std::default::Default;
-use std::io::{self, Write};
+use std::io;
 use ykpack;
 
-// Duplicates of LLVM types defined elsewhere, copied to avoid cyclic dependencies. Whereas the
-// LLVM backend expresses pointers to these using references, we use raw pointers so to as avoid
-// introducing lifetime parameters to the SirCx (and thus into TyCtxt and every place that uses
-// it).
-extern "C" {
-    pub type Value;
-}
-extern "C" {
-    pub type BasicBlock;
-}
-extern "C" {
-    pub type Builder;
+/// A collection of in-memory SIR data structures to be serialised.
+/// Each codegen unit builds one instance of this which is then merged into a "global" instance
+/// when the unit completes.
+#[derive(Default)]
+pub struct Sir {
+    pub funcs: RefCell<Vec<ykpack::Body>>,
 }
 
-newtype_index! {
-    pub struct SirFuncIdx {
-        DEBUG_FORMAT = "SirFuncIdx({})"
+impl Sir {
+    /// Returns `true` if we should collect SIR for the current crate.
+    pub fn is_required(tcx: TyCtxt<'_>) -> bool {
+        (tcx.sess.opts.cg.tracer.encode_sir()
+            || tcx.sess.opts.output_types.contains_key(&OutputType::YkSir))
+            && tcx.crate_name(LOCAL_CRATE).as_str() != "build_script_build"
```

Comment:
Should `build_script_build` be a constant?

Reply:
Arguably upstream should have defined one, but (last time I checked) haven't.  We can add one for our stuff if you like, but I don't want to make changes to other parts of the compiler.

Thoughts?

Reply:
That string doesn't appear in the main Rust repo; anyway, it's not a big deal either way.

Reply:
I don't mind defining a constant tbh.